### PR TITLE
ENH: Add decrypt support for V5 and AES-128, AES-256 (R5 only)

### PR DESCRIPTION
**Pull request by exiledkingcc**
rewrite the encryption part to support V4 and AES-128 encryption (ONLY decrypt for now).

i like the idea of [PyPDF2 cleanup](https://github.com/py-pdf/PyPDF2/issues/658), so this is Python3 ONLY.

this commit needs PyCryptodome for AES operations.

the encrypt part will be added some time later, and maybe AES-256 support will be added too, if it's not difficult.

### CODE REVIEW

**Comment by exiledkingcc** [comment]
local tox test is OK with py38.
it seems github failed because dependency PyCryptodome not installed.
does it not read tox.ini?

**Comment by MartinThoma** [comment]
> does it not read tox.ini?

CI (Github Actions) is defined [here](https://github.com/py-pdf/PyPDF2/blob/main/.github/workflows/github-ci.yaml). It uses [requirements/ci.txt](https://github.com/py-pdf/PyPDF2/blob/main/requirements/ci.txt) - and manually entered packages for Python 2.7

**Comment by MartinThoma** [comment]
As this PR is rather big, I would add it in the 2.0 release or later. Simply to ensure we can leave the 1.X parts soon-ish. Then you also don't have to care about the 2.7 support as it will be dropped with PyPDF2 version 2.0

**Comment by exiledkingcc** [comment]
encryption R=6 is used by PDF 2.0.
i can't find a "PDF 2.0 specification" on web, so it's not supported

File: PyPDF2/encryption.py [hunk 1]
```diff
@@ -0,0 +1,721 @@
...
+class CryptBase:
+    def encrypt(self, data: bytes) -> bytes: return data
+    def decrypt(self, data: bytes) -> bytes: return data
+
+
+class CryptIdentity(CryptBase): pass
+
+
+try:
+    from Crypto.Cipher import ARC4, AES
```
Comment (MasterOdin) [inline PyPDF2/encryption.py:47]: What's the motivation to have it include both attempt to use https://github.com/Legrandin/pycryptodome and also a fallback to an embedded method? Feels like should either add dependencies _or_ have decryption be included, but having both seems like added maintenance for little gain.

Comment (exiledkingcc) [reply PyPDF2/encryption.py:47]: at first, I used PyCryptodome, then I  found PyPDF2 has no dependencies at all, so I try not to introduce a dependency by adding AES code in PyPDF2.

encryption/decryption is not always needed for PDF, it's better to make PyCryptodome optional for users who do not need it.
for some other users, they may prefer PyCryptodome for better performance.
so I provide both.

maybe there should be a discussion for this?

Comment (MasterOdin) [reply PyPDF2/encryption.py:47]: For PDF encryption / decryption in the wild, it's really just using AES out in the wild, right? So it's not like if to have in-house cryptography around this, the code for that wouldn't be too large, and should be relatively stable?

Comment (exiledkingcc) [reply PyPDF2/encryption.py:47]: PDF specification also defined Public Key algorithms for PDF encryption, but I never see such PDF files.
most encrypted PDF files use only RC4 and AES, and some hash functions.
AES code can be very stable.

Comment (MartinThoma) [reply PyPDF2/encryption.py:47]: Thinking about it, I really would like to avoid having to maintain security-critical parts. I worry more about the encryption part than the decryption part.

If we mess up encryption, users might end up less secure then they expect. If we mess up decryption, users will just see an error.

Maybe we can have "inline imports" (importing Crypto within the encrypt / decrypt function). So we could make PyCryptodome an optional dependency that people would need to install if they want to use encryption / decryption. Maybe we could also get rid of a part of the current codebase this way?

To clarify: I don't see that I will properly maintain the crypto parts. This is the reason why I have this tendency. What do you think?

Comment (exiledkingcc) [reply PyPDF2/encryption.py:47]: inline imports is a good choice, actually it's what I did at first.
when I added AES code, I try to make the lib independent, but ignored the burden of maintaining.
I agree with that it's better to leave the AES to PyCryptodome when users need it.
I will remove _aes.py later.

**Comment by MartinThoma** [comment]
Very nice, thank you :hugs:

**Comment by MartinThoma** [comment]
One adjustment is still necessary: In `setup.cfg` needs to be a section:

```
[options.extras_require]
crypto = PyCryptodome
```

So people can install it with `pip install PyPDF2[crypto]`

Is there a minimum version of PyCryptodome this PR needs?

**Comment by exiledkingcc** [comment]
> One adjustment is still necessary: In `setup.cfg` needs to be a section:
> 
> ```
> [options.extras_require]
> crypto = PyCryptodome
> ```
> 
> So people can install it with `pip install PyPDF2[crypto]`

good idea!

> Is there a minimum version of PyCryptodome this PR needs?

no

**Comment by MartinThoma** [comment]
Nice! From my perspective this looks ready to be merged. However, due to the missing 2.7 support, it will take a while (until the PyPDF2 2.0 release).

I hope I can make that release on 1st of July. I really want to get this done: https://github.com/py-pdf/PyPDF2/discussions/752

If I don't get any comments on [what should be part of PyPDF2 1.x?](https://github.com/py-pdf/PyPDF2/discussions/753), I will start working on 2.0 on 1st of May.

**Comment by MartinThoma** [comment]
I think I'll add a "no-python27" marker ... on the other hand, if nobody reacts to https://github.com/py-pdf/PyPDF2/discussions/753 I will start working on the PyPDF2 2.0 release from 1st of May :smile:

**Comment by MartinThoma** [comment]
I still want to go through https://github.com/py-pdf/PyPDF2/pull/817 and merge 
* https://github.com/py-pdf/PyPDF2/pull/799 
* https://github.com/py-pdf/PyPDF2/pull/794
* https://github.com/py-pdf/PyPDF2/pull/788
Plus maybe add [some](https://github.com/py-pdf/PyPDF2/discussions/798) and maybe add some magic methods / camel_case method names + deprecation warnings for snakeCase method names: https://github.com/py-pdf/PyPDF2/discussions/751

**Comment by exiledkingcc** [comment]
keep update with main branch

**Comment by MartinThoma** [comment]
Just a heads-up: It might take all of May to finish https://github.com/py-pdf/PyPDF2/pull/859 . Especially the type annotations still take some time.

**Comment by MartinThoma** [comment]
We finally did it :-) the current main branch is Python 3. 6+🎉🥳

would you be so kind to update the pr?

**Comment by exiledkingcc** [comment]
keep update with main branch.

**Comment by MartinThoma** [comment]
Thank you for updating the PR :pray: 

Would you please add the following at the end of the `setup.cfg` file:

```
[options.extras_require]
crypto = pycryptodome
```

This means that people can `pip install PyPDF2[crypto]` which will also add pycryptodome.

File: PyPDF2/_reader.py [hunk 2]
```diff
@@ -1058,16 +1051,11 @@ def get_object(self, indirect_reference: IndirectObject) -> Optional[PdfObject]:
             # override encryption is used for the /Encrypt dictionary
             if not self._override_encryption and self.is_encrypted:
                 # if we don't have the encryption key:
-                if not hasattr(self, "_decryption_key"):
+                if not hasattr(self, "_encryption"):
```
Comment (MartinThoma) [inline PyPDF2/_reader.py:1054]: Why did you change the name?

File: PyPDF2/_reader.py [hunk 3]
```diff
@@ -1081,46 +1069,6 @@ def get_object(self, indirect_reference: IndirectObject) -> Optional[PdfObject]:
...
-    def getObject(
-        self, indirectReference: IndirectObject
-    ) -> Optional[PdfObject]:  # pragma: no cover
-        """
-        .. deprecated:: 1.28.0
-
-            Use :meth:`get_object` instead.
-        """
-        deprecate_with_replacement("getObject", "get_object")
-        return self.get_object(indirectReference)
```
Comment (MartinThoma) [inline PyPDF2/_reader.py:1093]: Please don't remove this

Comment (exiledkingcc) [reply PyPDF2/_reader.py:1093]: a merge mistake. it will be restored.

File: PyPDF2/_reader.py (hunk 2 above)
Comment (exiledkingcc) [reply PyPDF2/_reader.py:1054]: it's not just the key, it's the object representing all the encrypt and decrypt staff.

**Comment by exiledkingcc** [comment]
updated again.
`getObject` is restored,
`setup.cfg` already has the extra section.

**Comment by MartinThoma** [comment]
I plant to merge this on 19th on June and create `PyPDF2==2.3.0`. I just want to give people a bit of time to report issues with 2.2.0 :-) 

So far it looks good to me and I haven't heard/seen anybody speaking against it

**Comment by MartinThoma** [comment]
As a release message, I suggest:

```
ENH: Add decrypt support for V5 and AES-128, AES-256 (R5 only)

This is a rewrite of the encryption part to support V4 and AES-128 encryption (ONLY decrypt for now)
PyCryptodome was added as an optional dependency for AES.

It does NOT add support for encryption R=6 as introduced by PDF 2.0

Closes #528 
```

What do you think about it @exiledkingcc ? Would you add/modify anything?

**Comment by exiledkingcc** [comment]
it seems OK for me, no change is needed.

File: PyPDF2/_reader.py [hunk 4]
```diff
@@ -353,6 +343,9 @@ def _get_num_pages(self) -> int:
                 self._override_encryption = True
                 self.decrypt("")
                 return self.trailer[TK.ROOT]["/Pages"]["/Count"]  # type: ignore
+            except DependencyError as e:
+                # make dependency error clear to users
+                raise e
```
Comment (MartinThoma) [inline PyPDF2/_reader.py:348]: Is that part necessary? What does it do? It seems to me that you just catch an exception and raise exactly the same exception again.... hence removing this block would not change the behavior?

File: PyPDF2/encryption.py [hunk 5]
```diff
@@ -0,0 +1,740 @@
...
+    DictionaryObject,
+    StreamObject,
+    TextStringObject,
+    createStringObject,
+
+)
+
+
+class CryptBase:
+    def encrypt(self, data: bytes) -> bytes:
```
Comment (MartinThoma) [inline PyPDF2/encryption.py:55]:
```suggestion
    def encrypt(self, data: bytes) -> bytes:  # pragma: no cover
```

File: PyPDF2/encryption.py [hunk 6]
```diff
@@ -0,0 +1,740 @@
...
+        return data
+
+    def decrypt(self, data: bytes) -> bytes:
```
Comment (MartinThoma) [inline PyPDF2/encryption.py:58]:
```suggestion
    def decrypt(self, data: bytes) -> bytes:  # pragma: no cover
```

File: PyPDF2/_reader.py (hunk 4 above)
Comment (exiledkingcc) [reply PyPDF2/_reader.py:348]: if your PDF need AES algorithm to decrypt, but you don't install pycrytodome, without this two line,you will just got "the pdf is not decrypted", with this, you will know that you just need install the dependcy.

Comment (exiledkingcc) [reply PyPDF2/_reader.py:348]: if your PDF need AES algorithm to decrypt, but you don't install pycrytodome, without this two line,you will just got "the pdf is not decrypted", with this, you will know that you just need install the dependcy.

Comment (MartinThoma) [reply PyPDF2/_reader.py:348]: Oh, interesting! Took me a minute to understand - thank you for pointing out that I need to look at the following two lines!

**Comment by xilopaint** [comment]
Is there any encrypted PDF file available that I could use for testing this PR? I mean some file that I couldn't decrypt before and now it's possible.

**Comment by pubpub-zz** [comment]
@xilopaint, test samples are included in this PR for autotest

**Comment by MartinThoma** [comment]
@xilopaint You can also easily create them with qpdf:

```
$ qpdf --encrypt foo bar 256 --force-R5 -- resources/crazyones.pdf crazyones-256.pdf
```

and then run

```python
from PyPDF2 import PdfReader

reader = PdfReader("crazyones-256.pdf", password="foo")
print(reader.pages[0].extract_text())
```

If you checkout `2.2.1` you will get:

```
Traceback (most recent call last):
  File "foo.py", line 3, in <module>
    reader = PdfReader("crazyones-256.pdf", password="foo")
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_reader.py", line 262, in __init__
    if password is not None and self.decrypt(password) == 0:
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_reader.py", line 1610, in decrypt
    return self._decrypt(password)
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_reader.py", line 1651, in _decrypt
    f"only algorithm code 1 and 2 are supported. This PDF uses code {encrypt_v}"
NotImplementedError: only algorithm code 1 and 2 are supported. This PDF uses code 5
```

**Comment by xilopaint** [comment]
@MartinThoma the v2.3.0 is broken for me. `_codecs` is not being installed by `pip` and I'm getting `ModuleNotFoundError: No module named 'PyPDF2._codecs'`. To make it work I had to copy the `_codecs` folder.

Also, for using the new enhanced decryption I had to install `pycryptodome` because it was not included as a dependency.

**Comment by MartinThoma** [comment]
@xilopaint Thank you for pointing it out - I just fixed it.

`pycryptodome` is an optional dependency which you can install via `pip install PyPDF2[crypto]` (depending on your shell, you might need to escape the `[` and `]`).

**Comment by xilopaint** [comment]
> (depending on your shell, you might need to escape the `[` and `]`)

Yeah, I had tried `pip install pypdf2[crypto]` before and it didn't work in my zsh shell. Now using double quotes it does work.

What's the reason for `pycryptodome` being an optional dependency? Just making the package more lightweight for people who don't need the enhanced encryption?

**Comment by MartinThoma** [comment]
> What's the reason for pycryptodome being an optional dependency? Just making the package more lightweight for people who don't need the enhanced encryption?

Yes exactly! Quite a lot of users don't need any encryption / decryption capabilities.

Additionally, pycryptodome is not a pure-python dependency. That means if it was a non-optional dependency, it might make it for some cases way harder / impossible to install / use

**Comment by xilopaint** [comment]
@MartinThoma my tests are failing in GitHub Actions since I've installed `pypdf2[crypto]`. Look:

https://github.com/xilopaint/alfred-pdf-tools/runs/6956399382?check_suite_focus=true

Although they're passing locally with no errors. Could you help me out?

**Comment by MartinThoma** [comment]
I just had a quick glance, but it seems like a Linux kernel module is missing. That is exactly the kind of reason why I wanted this dependency to be optional.

If you don't install the crypto part, the pure-python implementation is used. That still offers encryption / decryption, but only older algorithms

**Comment by xilopaint** [comment]
Hey @MartinThoma, what does R5 mean in the context of cryptography?

**Comment by MartinThoma** [comment]
That is a PDF specific thing. It's short for "revision". You can read more in the specs in the "Standard Encryption Dictionary" chapter.

image

**Comment by xilopaint** [comment]
Thanks!
